torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp

```diff
@@ -0,0 +1,350 @@
+#include <torch/csrc/jit/passes/fixup_trace_scope_blocks.h>
+
+#include <torch/csrc/jit/passes/canonicalize.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+#include <torch/csrc/jit/passes/inliner.h>
+#include <torch/csrc/jit/passes/lower_tuples.h>
+#include <torch/csrc/jit/script/schema_matching.h>
+
+#include <algorithm>
+
+namespace torch {
+namespace jit {
+
+using BlockInfoMap = std::unordered_map<Block*, int>;
+
+void inspectBlocks(Block* b, BlockInfoMap& block_info, int level = 0) {
+  TORCH_INTERNAL_ASSERT(!block_info.count(b));
+  block_info[b] = level;
+  for (Node* n : b->nodes()) {
+    if (n->kind() == prim::FakeScopeBlock) {
+      inspectBlocks(n->blocks()[0], block_info, level + 1);
+    }
+  }
+}
+
+struct RemappedValueInfo {
+  Value* v;
+  int level;
+};
+using RemappingTable = std::unordered_map<Value*, RemappedValueInfo>;
+
+void makeDefsDominateUses(
+    Block* b,
+    RemappingTable& remap,
+    const BlockInfoMap& block_info) {
+  auto find_common_ancestor_block_level = [&](Node* n) {
```

Comment:
I think this should also be an API on Node, and can be more easily implemented when the level of each node is known pseudo code:

`while (one level is deeper than another) { go up one level }; while (block_lhs != block_rhs) { make both nodes go up a level; }